OutSystems/outsystems-datagrid-reactive
# RGRIDT-652: [BUG] Search with API fix Regex and logic

This PR is for RGRIDT-652 - Search with API is trying to match pair key and value instead of just the value

### What was happening
* 1- Searching for column headers shows all rows. Searching for single characters or words that are also contained in the column headers, will retrieve all rows.
![Problem1](https://user-images.githubusercontent.com/6432232/109172543-cf25ab80-777a-11eb-8200-23472ee8c3b7.gif)

* 2- **[NEW]** When using custom columns searching for values that exist but only on columns that are not being shown by the grid (this happens when the grid is using custom columns), the search mechanism will retrieve those rows as well.
![Problem2](https://user-images.githubusercontent.com/6432232/109172985-3f343180-777b-11eb-9c7e-8e5e778608a4.gif)

### What was done
* To solve **(1)** (First attempt was replaced by the second solution):
It was solved by replacing the following line from the Search method:
`const rx = new RegExp(searchedValue,'i');`
To a more explicit regular expression that only searches on the values of the stringified dataItem.
`const rx = new RegExp('([:]\s*")[^"]('+searchedValue+')[^"]*', 'i');`

So where the grid was previously retrieving a **True**  for matching the char “**F**” inside the stringified dataItem on row id 2:
{"Employee":{"Id":2,"Employee_Name":"Ait Sidi, Karthikeyan ","Married":true,"Salary":104437,"Position":"Sr. DBA","Sex":"M","Dateo`f`Hire":"2015-03-30T00:00:00.000Z","Department":"IT/IS"}}, 
we are now retriving null after trying to match.

* To solve **(2)**:
The solution turned out to be an extension of the previous solution. We have to check if the word exists on any of the rows, and in the case of existing, we have to check the name of the column where the searched value got the match.
With the help of @bmartinho we analyzed the way Wijmo is currently doing the search by client-side and for a better inspection, I created a sketch with both, OutSystems and Wijmo ways of filtering the data by a keyword/text.
![aww-board](https://user-images.githubusercontent.com/6432232/109173064-583ce280-777b-11eb-9fc1-1d5d54a0a38a.jpg)

* Added a new input parameter (not mandatory, False by default) to SearchData Block:
![image](https://user-images.githubusercontent.com/6432232/109176572-de0e5d00-777e-11eb-935c-81177761d8c4.png)
The goal of this new parameter is to allow searching on columns that are hidden or are set to not visible on the configuration of the column (in case you are using custom columns). It is now being passed into the API on our code.


### Logic and the brainstorming:
After comparing both options, we managed to understand that if we search for a keyword that only exists on one row, with the OS way, the filter will go through all dataItems and check if the keyword exists inside the (stringified) dataItem with Regex. And will only continue in that specific case. 
But using Wijmo’s way, what they do is converting all dataItems into an array (that at the end is joined into a string) and this one contains the values for the columns that exist. So the cycle of checking if the column exists or is visible will run for all rows no matter what, duplicating the arrays. At the end, it will check one by one with regex if the keyword is inside that array (which is a string by now).
In the end, I followed with option 1, just like to option name suggests - OS way.

### Test Steps
On a grid with a Search block
1. Search for any text that is equal to a column header.
Expected: No rows should appear on the grid (unless they have the column header written on any cell value)

On a grid with custom columns and a Search block with allowSearchHiddenColumns = True
1. Hide all textual columns
2. Search for "pink"
Expected: Only the rows with the text "pink" should appear on the grid.

On a grid with custom columns and a Search block with allowSearchHiddenColumns = False
1. Hide all textual columns
2. Search for "pink"
Expected: No rows should appear on the grid.

On a grid with autogenerated columns and a Search block with allowSearchHiddenColumns = True
1. Hide all textual columns
2. Search for "pink"
Expected: Only the rows with the text "pink" should appear on the grid.

On a grid with autogenerated columns and a Search block with allowSearchHiddenColumns = False
1. Hide all textual columns
2. Search for "pink"
Expected: No rows should appear on the grid.


### Screenshots
**After the fix:**
(problem 1)
![gitsolution3](https://user-images.githubusercontent.com/6432232/109174341-ab636500-777c-11eb-8c44-b6eefe99e5b6.gif)

(problem 2)
![gitsolution](https://user-images.githubusercontent.com/6432232/109174402-bb7b4480-777c-11eb-8339-80e5ed5914bd.gif)
![gitsolution2](https://user-images.githubusercontent.com/6432232/109174415-bf0ecb80-777c-11eb-9468-93b7dfea8e5b.gif)

### Checklist
* [x] requires changes in OutSystems (if so, provide a module with changes) - **RGRIDT-652**
